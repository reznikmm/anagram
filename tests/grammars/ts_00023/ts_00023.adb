--  Check conflict resolution

with Gela.Grammars;
with Gela.Grammars.Conflicts;
with Gela.Grammars.Constructors;
with Gela.Grammars_Debug;
with Gela.Grammars.Reader;
with Gela.Grammars.LR_Tables;
with Gela.Grammars.LR.Simple;

with Ada.Text_IO;

procedure TS_00023 is

   procedure Print_Action
     (Table : Gela.Grammars.LR_Tables.Table;
      State : Gela.Grammars.LR.State_Index;
      T     : Gela.Grammars.Terminal_Count);

   ------------------
   -- Print_Action --
   ------------------

   procedure Print_Action
     (Table : Gela.Grammars.LR_Tables.Table;
      State : Gela.Grammars.LR.State_Index;
      T     : Gela.Grammars.Terminal_Count)
   is
      use Gela.Grammars.LR_Tables;
      use type Gela.Grammars.Terminal_Count;
      use type Gela.Grammars.LR.State_Count;

      S : constant Gela.Grammars.LR.State_Count :=
        Shift (Table, State, T);
      R : constant Reduce_Iterator := Reduce (Table, State, T);
   begin
      Ada.Text_IO.Put (' ');

      if T = 0  and then Finish (Table, State) then
         Ada.Text_IO.Put ("FINISH   ");
      elsif S /= 0 then
         Ada.Text_IO.Put ("SHIFT ");
         Ada.Text_IO.Put (Gela.Grammars.LR.State_Count'Image (S));

         if S in 1 .. 9 then
            Ada.Text_IO.Put (' ');
         end if;
      elsif not Is_Empty (R) then
         Ada.Text_IO.Put ("REDUCE");

         Ada.Text_IO.Put
           (Gela.Grammars.Production_Index'Image (Production (R)));

         if Production (R) in 1 .. 9 then
            Ada.Text_IO.Put (' ');
         end if;
      else
         Ada.Text_IO.Put ("Error    ");
      end if;
   end Print_Action;

   G : constant Gela.Grammars.Grammar := Gela.Grammars.Reader.Read ("test.ag");
   AG : constant Gela.Grammars.Grammar :=
     Gela.Grammars.Constructors.To_Augmented (G);
   use Gela.Grammars;
begin
   --  Gela.Grammars_Debug.Print (AG);

   declare
      use Gela.Grammars.LR;

      Table : LR_Tables.Table_Access := Simple.Build (AG);
      R : Gela.Grammars.Conflicts.Resolver;
   begin
      R.Resolve (AG, Table.all);
      Gela.Grammars_Debug.Print_Conflicts (AG, Table.all);

      for S in 1 .. LR_Tables.Last_State (Table.all) loop
         Ada.Text_IO.Put (State_Index'Image (S));

         if S <= 9 then
            Ada.Text_IO.Put (' ');
         end if;

         for T in 0 .. AG.Last_Terminal loop
            Print_Action (Table.all, S, T);
         end loop;

         Ada.Text_IO.Put (ASCII.HT);

         for NT in 1 .. AG.Last_Non_Terminal loop
            Ada.Text_IO.Put
              (State_Count'Image (LR_Tables.Shift (Table.all, S, NT)));
         end loop;

         Ada.Text_IO.New_Line;
      end loop;

      LR_Tables.Free (Table);
   end;
end TS_00023;
