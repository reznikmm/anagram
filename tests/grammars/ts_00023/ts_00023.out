 1  Error     SHIFT  3  Error     Error     SHIFT  2  Error    	 4 0
 2  Error     SHIFT  3  Error     Error     SHIFT  2  Error    	 5 0
 3  REDUCE 4  Error     REDUCE 4  REDUCE 4  Error     REDUCE 4 	 0 0
 4  FINISH    Error     SHIFT  7  SHIFT  6  Error     Error    	 0 0
 5  Error     Error     SHIFT  7  SHIFT  6  Error     SHIFT  8 	 0 0
 6  Error     SHIFT  3  Error     Error     SHIFT  2  Error    	 9 0
 7  Error     SHIFT  3  Error     Error     SHIFT  2  Error    	 10 0
 8  REDUCE 3  Error     REDUCE 3  REDUCE 3  Error     REDUCE 3 	 0 0
 9  REDUCE 2  Error     REDUCE 2  REDUCE 2  Error     REDUCE 2 	 0 0
 10 REDUCE 1  Error     REDUCE 1  SHIFT  6  Error     REDUCE 1 	 0 0
